### PR TITLE
Chunking  in StoreLargoSession

### LINKED ISSUE #111: Chunking required in StoreLargoSession
The StoreLargoSession request validates if all [image annotations](https://github.com/biigle/largo/blob/d947ddd63e1bc7853e84ebaf3dd250b6508c9b6e/src/Http/Requests/StoreLargoSession.php#L113-L119) and [video annotations](https://github.com/biigle/largo/blob/d947ddd63e1bc7853e84ebaf3dd250b6508c9b6e/src/Http/Requests/StoreLargoSession.php#L129-L135) belong to the specified volume. There was a request that tried to store more than 65535 annotations, which is the parameter limit of the SQL query, so the query failed. We need to chunk the annotation array and perform multiple queries if the number of annotations gets too large (e.g. 65000).

### PR DESCRIPTION
**Pull request by aneust**
Chunk image and video annotations if the number of annotations gets too large (65000)

- Updated methods to handle large sets of annotations by splitting them into chunks of 65,000.
- Done for project and volume controller.

Resolves #111

### CODE REVIEW

**Review by mzur — CHANGES_REQUESTED** [review]
Some tests would highlight the issues I pointed out. Since it is difficult to test this (without actually creating 65k annotations) we should add the following: In `./config/biigle.php` we should add a new config parameter `db_param_limit => 65000` with an appropriate comment. In the request classes you can use this number with `config('biigle.db_param_limit')`.

Then in the tests you can easily test the new behavior if you lower the limit with `config(['biigle.db_param_limit' => 1])`. With this you only have to create two annotations and they should already be chunked.

In a [follow-up issue](https://github.com/biigle/core/issues/941) we could implement usage of the new config value in other parts of BIIGLE that already do chunking based on the parameter limit. Then we can add tests there, too.

Edit: You have to create a new branch and PR in biigle/core for the new config value.

**Comment by aneust** [comment]
@mzur : I am not sure if I should set the db_param_limit to 65535 or 65534, because chunking works with the latter.

File: src/Http/Requests/StoreLargoSession.php [hunk 1]
```diff
@@ -112,23 +112,18 @@ protected function getAffectedAnnotations($dismissed, $changed)
      */
     protected function imageAnotationsBelongToVolumes($annotations, $volumes)
     {
-        $leni = count($annotations);
-        if ($leni<65000){
-            return !ImageAnnotation::join('images', 'image_annotations.image_id', '=', 'images.id')
-                ->whereIn('image_annotations.id', $annotations)
+        $chunkedI = array_chunk($annotations,config('biigle.db_param_limit')-1);
```
Comment (mzur) [inline src/Http/Requests/StoreLargoSession.php:115]: The annotation IDs are not the only parameters in this query. There are also the volume IDs. So the chunk size should be like : `$paramLimit - count($volumes) -  1`.

File: src/Http/Requests/StoreLargoSession.php [hunk 2]
```diff
@@ -141,22 +136,17 @@ protected function imageAnotationsBelongToVolumes($annotations, $volumes)
...
-            return !VideoAnnotation::join('videos', 'video_annotations.video_id', '=', 'videos.id')
-                ->whereIn('video_annotations.id', $annotations)
-                ->whereNotIn('videos.volume_id', $volumes)
-                ->exists();
-        }
-        else {
-            $chunkedV = array_chunk($annotations,65000);
-            foreach($chunkedV as $cv){
-                return !VideoAnnotation::join('videos', 'video_annotations.video_id', '=', 'videos.id')
+        $chunkedV = array_chunk($annotations,config('biigle.db_param_limit')-1); // config('biigle.db_param_limit')
```
Comment (mzur) [inline src/Http/Requests/StoreLargoSession.php:139]:
```suggestion
        $chunkedV = array_chunk($annotations,config('biigle.db_param_limit')-1);
```

Comment (mzur) [inline src/Http/Requests/StoreLargoSession.php:139]: Same as above, you have to consider the volume IDs, too.

File: tests/Http/Controllers/Api/Projects/LargoControllerTest.php [hunk 3]
```diff
@@ -443,4 +446,64 @@ public function testStoreJobStillRunningVideoAnnotations()
...
             ])->assertStatus(422);
     }
+
+    public function testChunkingImgProject()
+    {
+        config(['biigle.db_param_limit' => 2]);
+        $image = ImageTest::create(['volume_id' => $this->imageVolume->id,
+                                    'filename'=>"testImage"]);
+
+        $imageAnnotations = ImageAnnotation::factory()->count(5)->create(['image_id' => $image->id]); // 70000
```
Comment (mzur) [inline tests/Http/Controllers/Api/Projects/LargoControllerTest.php:456]:
```suggestion
        $imageAnnotations = ImageAnnotation::factory()->count(5)->create(['image_id' => $image->id]);
```

File: tests/Http/Controllers/Api/Projects/LargoControllerTest.php [hunk 4]
```diff
@@ -443,4 +446,64 @@ public function testStoreJobStillRunningVideoAnnotations()
...
+        $label = Label::factory()->create(); 
+
+
+        $imageAnnotations->each(function($imageAnnotation) use($label){
+            ImageAnnotationLabelTest::create([
+                'annotation_id' => $imageAnnotation->id,
+                'user_id' => $this->editor()->id,
+                'label_id' => $label->id,
+            ]);
+        });
```
Comment (mzur) [inline tests/Http/Controllers/Api/Projects/LargoControllerTest.php:466]: PHP hint: There are also arrow functions which look a little cleaner in cases like these.

```suggestion
        $imageAnnotations->each(fn ($imageAnnotation) => ImageAnnotationLabelTest::create([
            'annotation_id' => $imageAnnotation->id,
            'user_id' => $this->editor()->id,
            'label_id' => $label->id,
        ]));
```

**Review by mzur — COMMENTED** [review]
Please use a couple more spaces for better readability. Example:

```diff
- $chunkedI = array_chunk($annotations,config('biigle.db_param_limit')-1);
+ $chunkedI = array_chunk($annotations, config('biigle.db_param_limit') - 1);
```

This is just an example, there are more locations where spaces would be good.

Also I was wondering what the `I` and `V` sufixes in the variable names should mean. "Image" and "Video"? Since these are handled in different methods, you don't really need a distinction in the name. But if you want to keep them, I want the variable names to be more descriptive (e.g. `chunkedImageAnnotations`). I'd prefer a shorter `chunkedAnnotations`, though :wink:

File: src/Http/Requests/StoreLargoSession.php [hunk 5]
```diff
@@ -112,13 +112,13 @@ protected function getAffectedAnnotations($dismissed, $changed)
      */
     protected function imageAnotationsBelongToVolumes($annotations, $volumes)
     {
-        $chunkedI = array_chunk($annotations,config('biigle.db_param_limit')-1);
```
Comment (mzur) [inline src/Http/Requests/StoreLargoSession.php]: Why did you remove the `-1` everywhere?

Comment (aneust) [reply src/Http/Requests/StoreLargoSession.php]: Using the volume ID count to set the chunk size resolved the error I had previously addressed with a value of -1. So there is no longer a need for it.

**Review by mzur — APPROVED** [review]
Thanks!
